### ISSUE TITLE
Available balance

### ISSUE THREAD

**Issue by Eknir** [ethersphere/swarm#1884]
`AvailableBalance` is defined as the balance of a node operator's chequebook, which is currently not promised to any party via issuance of a cheque. As such, it is always less or equal to the `liquidBalance` of a chequebook.

The availableBalance can be calculated by reading the event-logs of a chequebook. From the events `Deposit` and `Withdraw`, we calculate `netDeposit` (sum of all amounts ever deposited - sum of all amounts ever withdrawn). From this, you subtract the sum of all `lastSentCheques`. This metric can be calculated differently as well, but I think the way described above is the neatest.

This metric is a valuable metric for a node operator, as it allows him to see how much money he still has `in stock`, hence: how long he can still operate without refunding.

This metric is valuable for the swap-package, as the package should not allow sending any more cheques if `availableBalance` is less than the `paymentThreshold`.

I propose to create a function that will return this metric and expose this function as well via the SwapAPI. 

Note: if the node ever sends cheques outside of the `swap` protocol, the metric is not correct anymore. There is no way to safeguard against this, but we should warn the node operator to never connect one chequebook to different Swarm instances, or use a chequebook used in a Swarm process to send cheques outside of this process.

**Comment by mortelli** [comment]
> `AvailableBalance` is defined as the balance of a node operator's chequebook, which is currently not promised to any party via issuance of a cheque. As such, it is always less or equal to the `liquidBalance` of a chequebook.

Just to clarify (for economics newbies like me :sweat_smile:) could you make explicit the formulas for both the _available balance_ as well as the _liquid balance_? So that the difference is clear.

> The availableBalance can be calculated by reading the event-logs of a chequebook. From the events `Deposit` and `Withdraw`, we calculate `netDeposit` (sum of all amounts ever deposited - sum of all amounts ever withdrawn). From this, you subtract the sum of all `lastSentCheques`. This metric can be calculated differently as well, but I think the way described above is the neatest.

To make sure we agree on the implementation, could you please specify what the other ways of calculating this are?

> Note: if the node ever sends cheques outside of the `swap` protocol, the metric is not correct anymore. There is no way to safeguard against this, but we should warn the node operator to never connect one chequebook to different Swarm instances, or use a chequebook used in a Swarm process to send cheques outside of this process.

Is this still true for all of the ways of implementing this feature?

**Comment by Eknir** [comment]
> Just to clarify (for economics newbies like me sweat_smile) could you make explicit the formulas for both the available balance as well as the liquid balance? So that the difference is clear.

`liquidBalance` can be read entirely from the on-chain smart-contract and is defined as `balance` (amount of wei deposited in the chequebook) - `hardDeposits` (amount of wei under hard deposit) ([source code](https://github.com/ethersphere/swap-swear-and-swindle/blob/2aec7a67458dd8100e635576299ca3c544800698/contracts/SimpleSwap.sol#L65)). `availableBalance` (I coined this term myself in this PR) is in theory only known to the chequebook owner as it comprises the value of cheques which have been sent, but not yet presented to the on-chain smart-contract. The formula of `availableBalance` is `liquidBalance` - value of cheques send, but not yet presented.

> To make sure we agree on the implementation, could you please specify what the other ways of calculating this are?

`liquidBalance() + SUM(cashed(peer)) - SUM(p.getLastSentCheque().CumulativeAmount)`
Cashed(peer) is a call not yet defined in Swarm ([but it exists in sw3](https://github.com/ethersphere/swap-swear-and-swindle/blob/2aec7a67458dd8100e635576299ca3c544800698/contracts/SimpleSwap.sol#L42)). 

They are not entirely the same ([link to comment](https://github.com/ethersphere/swarm/pull/1892#discussion_r337364915)):
> Is this still true for all of the ways of implementing this feature?

There is no way to prevent this from happening (aside from making sure that when it happens and is detected, the node will be severely punished by peers). The trick lies in the fact that a cheque send is only known to the receiver and the sender of the cheque until it is presented to the on-chain smart-contract. If the sender is a different process than Swarm, there is no way of Swarm knowing about the fact that this cheque is sent. Such a behavior (using the chequebook with different processes) makes it possible that a cheque bounces (even if we have made the Swarm code such as to prevent cheques from bouncing) and should be regarded as fraud.
